Add CoreDisTools Library

swaroop-sridhar opened the pull request:
This change adds CoreDisTools -- a library containing
LLVM disassembly tools for CoreClr to LLILC repo.

Currently, the library supports instruction-wise
disassembly API required for implementing GcStress.

coredistools.h -- the export/import header
coredistools.cpp -- the implementation as a wrapper around
    LLVM disassemblers to suite the required API.

This change also adds code to package the library into a
NuGet package.

swaroop-sridhar commented:
@RussKeldorph please review.

RussKeldorph commented on include/CoreDisTools/coredistools.h:
```diff
@@ -0,0 +1,36 @@
+#define LLILC_TOOLS_COREDISTOOLS
+
+#if !defined(DllIface)
+#define DllIface extern "C" __declspec(dllimport)
```
I would expect this header to be compatible with C unless there is some convention discouraging that that I'm not aware of (possible).  Constructs like `extern "C"` need a `defined(__cplusplus)` guard.  Also, is `__declspec(dllimport)` relevant/appropriate on non-Win?

RussKeldorph commented on include/CoreDisTools/coredistools.h:
```diff
@@ -0,0 +1,36 @@
+#define DllIface extern "C" __declspec(dllimport)
+#endif // defined(DllIface)
+
+enum TargetArch {
```
Related to C-compat comment above.  I think this should be a `typedef`.

RussKeldorph commented on include/CoreDisTools/coredistools.h:
```diff
@@ -0,0 +1,36 @@
+#endif // defined(DllIface)
+
+enum TargetArch {
+  Target_Host,
```
I believe there's a subtle ambiguity with `Target_Host` on Linux/ARM since there both ARM and Thumb code can exist in the same process.  This may not matter on Windows since generally only Thumb is allowed, though the concept is not unheard of even on Windows since I believe even Windows mixes x86 and x64 under WoW.

One option would be to have a major/minor architecture parameter, similar to how LLVM classifies things, but I think it's probably simpler to just make Thumb a peer of ARM for our purposes.

NOTE that I'd also probably be ok with just ignoring this issue under the assumption that this library is purpose-built for CoreCLR, which doesn't acknowledge the possibility of more than one simultaneous target architecture.  In that case, what you've done makes perfect sense, though I would probably rename Target_Arm to Target_Thumb to allow for unambiguous future expansion.

RussKeldorph commented on include/CoreDisTools/coredistools.h:
```diff
@@ -0,0 +1,36 @@
+  Target_Arm64
+};
+
+class CorDisasm;
```
`struct` instead of `class` for C compat?

kyulee1 commented:
Can you fix the build break and squash the change? I will take a look then.

swaroop-sridhar commented:
@kyulee1: the build issue is now fixed. The remaining error is an unrelated llilc test timeout

kyulee1 commented on include/CoreDisTools/coredistools.h:
```diff
@@ -0,0 +1,52 @@
+};
+
+struct CorDisasm;
+DllIface CorDisasm *InitDisasm(enum TargetArch Target);
```
Can you make Target argument default as Target_Host so that we can skip setting it on the same  Host/Target?

swaroop-sridhar replied:
@kyulee1: We're not using the default parameters in the DLL interface in order to keep the header compatible with C. Please see @RussKeldorph's comment above.

kyulee1 replied:
Ok. That's fine without having a default value.

kyulee1 commented on lib/CoreDisTools/.nuget/toolchain.ubuntu.14.04-x64.Microsoft.DotNet.Coredistools.nuspec:
```diff
@@ -0,0 +1,21 @@
+  </metadata>
+  <files>
+    <file src="../libcoredistools.so" target="runtimes/ubuntu.14.04-x64/native/libcoredistools.so" />
+    <file src="../include/coredistools.h" target="runtimes/win7-x64/native/coredistools.h" />
```
Please fix the folder name "win7-x64" -> "ubuntu.14.04-x64".

kyulee1 commented on lib/CoreDisTools/.nuget/toolchain.osx.10.10-x64.Microsoft.DotNet.Coredistools.nuspec:
```diff
@@ -0,0 +1,21 @@
+  </metadata>
+  <files>
+    <file src="../libcoredistools.dylib" target="runtimes/osx.10.10-x64/native/libcoredistools.dylib" />
+    <file src="../include/coredistools.h" target="runtimes/win7-x64/native/coredistools.h" />
```
Please fix the folder name "win7-x64" -> "osx.10.10-x64".

kyulee1 commented on lib/CoreDisTools/coredistools.exports:
```diff
@@ -0,0 +1,3 @@
+InitDisasm
+FinishDisasm
+DisasmInstruction
```
It appears PrintInstruction is an export in the header file but you didn't specify it here.

kyulee1 commented on lib/CoreDisTools/coredistools.cpp:
```diff
@@ -0,0 +1,236 @@
+  return Disasm->disasmInstruction(Address, Bytes, Maxlength);
+}
+
+DllIface void PrintInstruction(const CorDisasm *Disasm, size_t Address,
```
PrintInstruction seems to do disasm and print instruction. Can we just add an optional argument to DisasmInstruction for this purpose instead of exporting it?

swaroop-sridhar replied:
We cannot use Optional arguments because of C compatibility. But I can add a mandatory argument to DisasmInstruction like: 

```
  DllIface size_t DisasmInstruction(const CorDisasm *Disasm, size_t Address,
                                                          const uint8_t *Bytes, size_t Maxlength, bool PrintAsm)
```

However I left it as two functions, because `DisasmInstruction` is designed for GcStress which is the sole user currently (to get the instruction size) and `PrintInstruction` is mostly used for debugging.

kyulee1 replied:
PrintInstruction functionally does DisasmInstruction and prints it, so I don't think it's much useful to have a separate API. Can you add the mandatory argument to DisasmInsutrction as you suggested while deleting PrintInstruction?

kyulee1 commented on lib/CoreDisTools/coredistools.cpp:
```diff
@@ -0,0 +1,236 @@
+  TargetArch TargetArch;
+  string TargetTriple;
+  const Target *TheTarget;
+  bool ShowInstructionBytes;
```
Is this always true? than we don't need to set this here.

kyulee1 commented on include/CoreDisTools/coredistools.h:
```diff
@@ -0,0 +1,52 @@
+                               const uint8_t *Bytes, size_t Maxlength);
+DllIface void FinishDisasm(const CorDisasm *Disasm);
+
+#endif // defined(LLILC_TOOLS_COREDISTOOLS)
```
Nit. The comment is about the negative definition -- !defined(LLILC_TOOLS_COREDISTOOLS).

kyulee1 commented on include/CMakeLists.txt:
```diff
@@ -1,5 +1,6 @@
 add_subdirectory(clr)
 add_subdirectory(Driver)
+add_subdirectory(CoreDisTools)
```
I know you're the one that adds this. But all of this additions including this CmakeLists.txt are useless. If you could clean up these since you're here, it would be great. Or at least let's not add this anymore including the empty CmakeLists.txt under CoreDisTools.

swaroop-sridhar commented:
Thanks for the review @kyulee1. I've addressed your comments in https://github.com/swaroop-sridhar/llilc/commit/bea7195c2be408f524239d5696c849e390dff39b.

kyulee1 commented on lib/CoreDisTools/.nuget/runtime.json:
```diff
@@ -0,0 +1,19 @@
+  "runtimes": {
+    "win7-x64": {
+      "Microsoft.DotNet.CoreDisTools": {
+        "toolchain.win7-x64.Microsoft.DotNet.Coredistools": "1.0.0-prerelease"
```
Case should be aligned for all filenames/ids: "Coredistools" -> "CoreDisTools".
Please make sure this in the other cases.

swaroop-sridhar replied:
Fixed, thanks.

kyulee1 commented:
@swaroop-sridhar The file names are still not updated --e.g,  lib/CoreDisTools/.nuget/toolchain.osx.10.10-x64.Microsoft.DotNet.Coredistools.nuspec ==> lib/CoreDisTools/.nuget/toolchain.osx.10.10-x64.Microsoft.DotNet.CoreDisTools.nuspec

swaroop-sridhar commented:
I've renamed the files in the .nuget directory. The .h, .cpp, .exports files are lower cased, following the convention in the other directories.

kyulee1 commented:
LGTM. Please squash change and fix formatting before merge.
